src/transport/grpc/mod.rs

```diff
@@ -0,0 +1,837 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! # Astarte GRPC Transport Module
+//!
+//! This module provides an implementation of the Astarte transport layer using the GRPC protocol.
+//! It defines the `Grpc` struct, which represents a GRPC connection, along with traits for publishing,
+//! receiving, and registering interfaces.
+
+pub mod convert;
+
+use std::{collections::HashMap, ops::Deref, sync::Arc};
+
+use astarte_message_hub_proto::{
+    astarte_data_type, message_hub_client::MessageHubClient, AstarteMessage, Node,
+};
+use async_trait::async_trait;
+use tokio::sync::Mutex;
+use uuid::Uuid;
+
+use crate::{
+    builder::{ConnectionConfig, DeviceBuilder},
+    interface::{
+        mapping::path::MappingPath,
+        reference::{MappingRef, ObjectRef},
+    },
+    interfaces::Interfaces,
+    shared::SharedDevice,
+    store::PropertyStore,
+    types::AstarteType,
+    validate::{ValidatedIndividual, ValidatedObject},
+    Interface, Timestamp,
+};
+
+use super::{Disconnect, Publish, Receive, ReceivedEvent, Register};
+
+use self::convert::map_values_to_astarte_type;
+
+/// Errors raised while using the [`Grpc`] transport
+#[derive(Debug, thiserror::Error)]
+pub enum GrpcTransportError {
+    #[error("Transport error while working with grpc: {0}")]
+    Transport(#[from] tonic::transport::Error),
+    #[error("Status error {0}")]
+    Status(#[from] tonic::Status),
+    #[error("Error while serializing the interfaces")]
+    InterfacesSerialization(#[from] serde_json::Error),
+    #[error("Attempting to deserialize individual message but got an object")]
+    DeserializationExpectedIndividual,
+    #[error("Attempting to deserialize object message but got an individual value")]
+    DeserializationExpectedObject,
+    #[error("Graceful close of the grpc channel failed, the Arc is still shared")]
+    GracefulClose,
+}
+
+/// Shared data of the grpc connection, this struct is internal to the [`Grpc`] connection
+/// where is wrapped in an arc to share an immutable reference across tasks.
+pub struct SharedGrpc {
+    client: Mutex<MessageHubClient<tonic::transport::Channel>>,
+    stream: Mutex<tonic::codec::Streaming<AstarteMessage>>,
+    uuid: Uuid,
+}
+
+/// This struct represents a GRPC connection handler for an Astarte device. It manages the
+/// interaction with the [astarte-message-hub](https://github.com/astarte-platform/astarte-message-hub), sending and receiving [`AstarteMessage`]
+/// following the Astarte message hub protocol.
+#[derive(Clone)]
+pub struct Grpc {
```

Comment:
I whould add a manual impl of debug on the inner shared fileds, or at least of the struct name, without locking

Reply:
Why not deriving the impl ? The derived impl should not lock anyways. And i will also derive `Debug` for the `SharedGrpc` object.

Reply:
It is actually pretty noisy
```
Grpc { shared: SharedGrpc { client: Mutex { data: MessageHubClient { inner: Grpc { inner: Channel, origin: /, compression_encoding: None, accept_compression_encodings: EnabledCompressionEncodings, max_decoding_message_size: None, max_encoding_message_size: None } } }, stream: Mutex { data: Streaming }, uuid: 67e55044-10b1-426f-9247-bb680e5fe0c8 } }
```
For now i just print `Grpc { uuid: 67e55044-10b1-426f-9247-bb680e5fe0c8, .. }`